crates/lib/src/execution_data.rs

```diff
@@ -18,3 +23,192 @@ pub struct ExecutionData {
     /// This will always be `None` for QVM execution.
     pub duration: Option<Duration>,
 }
+
+#[derive(Debug, Clone, PartialEq)]
+#[repr(transparent)]
+pub struct ReadoutMap(HashMap<MemoryReference, ReadoutValues>);
+
+impl ReadoutMap {
+    /// Given a known readout field name and index, return the result's ReadoutValues, if any.
+    pub fn get_readout_values(&self, field: &str, index: u64) -> Option<ReadoutValues> {
+        let readout_values = self.0.get(&MemoryReference {
+            name: String::from(field),
+            index,
+        })?;
+
+        Some(readout_values.clone())
+    }
+
+    /// Given a known readout field name, return the result's ReadoutValues for all indices, if any.
+    pub fn get_readout_values_for_field(
+        &self,
+        field: &str,
+    ) -> Result<Option<Vec<Option<ReadoutValues>>>, TryFromIntError> {
+        let mut readout_values = Vec::new();
+        for (memref, values) in &self.0 {
+            let MemoryReference { name, index } = memref;
+            let index = usize::try_from(*index)?;
+            if name == field {
+                if readout_values.len() <= index {
+                    readout_values.resize(index + 1, None);
+                }
+                readout_values[index] = Some(values.clone());
+            }
+        }
+
+        Ok(match readout_values.len() {
+            0 => None,
+            _ => Some(readout_values),
+        })
+    }
+
+    /// `readout_values` maps program-defined readout to result-defined readout, e.g.:
+    ///     { "ro[0]": "q0", "ro[1]": "q1" }
+    /// where `ro[0]` is defined in the original program, and `q0` is what comes back in execution results.
+    /// Here we map the result-defined readout values back to their original result-defined names.
+    pub(crate) fn from_mappings_and_values(
+        readout_mappings: &HashMap<String, String>,
+        readout_values: &HashMap<String, ReadoutValues>,
+    ) -> Result<Self, MemoryReferenceParseError> {
+        let result = readout_values
+            .iter()
+            .flat_map(|(readout_name, values)| {
+                readout_mappings
+                    .iter()
+                    .filter(|(_, program_alias)| *program_alias == readout_name)
+                    .map(|(program_name, _)| program_name.clone())
+                    .map(parse_readout_register)
+                    .map(|reference| Ok((reference?, values.clone())))
+                    .collect::<Result<Vec<_>, MemoryReferenceParseError>>()
```

Comment:
Shouldn't need to specify the error type.

```suggestion
                    .collect::<Result<Vec<_>, _>>()
```